src/python/pants/help/flag_error_help_printer.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import difflib
+
+from pants.engine.unions import UnionMembership
+from pants.help.help_info_extracter import HelpInfoExtracter
+from pants.help.maybe_color import MaybeColor
+from pants.option.errors import UnknownFlagsError
+from pants.option.options import Options
+from pants.option.scope import GLOBAL_SCOPE
+
+
+class FlagErrorHelpPrinter(MaybeColor):
+    """Prints help related to erroneous command-line flag specification to the console."""
+
+    def __init__(self, options: Options):
+        super().__init__(options.for_global_scope().colors)
+        self._bin_name = options.for_global_scope().pants_bin_name
+        self._all_help_info = HelpInfoExtracter.get_all_help_info(
+            options,
+            # We only care about the options-related help info, so we pass in
+            # dummy values for union_membership and consumed_scopes_mapper.
```

Comment:
It looks like we're actually creating dummy instances of each argument to this method. If we were to separate the help extraction process for options from unions and the scope mapper, we might avoid having to create any of these dummies.

I couldn't immediately see how to do that, because it's not quite clear what `ConsumedScopesMapper` is supposed to be doing. But it might be worth leaving a TODO to avoid creating a dummy for every argument of this method.

It would at least make it easier to follow the logic in the `handle_unknown_flags()` method if we could expose e.g. `self._all_help_info.scope_to_help_info.values()` and `self._all_help_info.scope_to_help_info[GLOBAL_SCOPE].collect_unscoped_flags()` as named methods of `AllHelpInfo` instead of reaching into those dicts here directly.

Reply:
Seems like more lifting than I want to do right now...